platform/security/idp/security-idp-client/src/main/java/org/codice/ddf/security/idp/client/IdpMetadata.java

```diff
@@ -223,8 +279,53 @@ public String getSigningCertificate() {
     return signingCertificate;
   }
 
+  @SuppressWarnings("unused")
   public String getEncryptionCertificate() {
     initCertificates();
     return encryptionCertificate;
   }
+
+  @Nullable
+  private EntityDescriptor extractRootEntityFromMap(Map<String, EntityDescriptor> edMap) {
+    Set<Map.Entry<String, EntityDescriptor>> entries = edMap.entrySet();
+    if (!entries.isEmpty()) {
+      return entries.iterator().next().getValue();
+    }
+    return null;
+  }
+
+  private class EntityData {
+    private final EntityDescriptor entityDescriptor;
+    private final Duration cacheDuration;
+    private final Instant validUntil;
+    private final Instant created;
+
+    EntityData(EntityDescriptor ed) {
+      entityDescriptor = ed;
+      if (getEntityDescriptor() == null) {
+        cacheDuration = null;
+        validUntil = null;
+        created = null;
+      } else {
+        created = Instant.now();
+        Long entityDuration = getEntityDescriptor().getCacheDuration();
+        DateTime entityValidity = getEntityDescriptor().getValidUntil();
+        this.cacheDuration = (entityDuration != null) ? Duration.ofMillis(entityDuration) : null;
+        this.validUntil = (entityValidity != null) ? entityValidity.toDate().toInstant() : null;
+      }
+    }
+
+    @Nullable
+    public EntityDescriptor getEntityDescriptor() {
+      return entityDescriptor;
+    }
+
+    private boolean isMetadataExpired() {
+      return !(cacheDuration == null || Instant.now().isAfter(created.plus(cacheDuration)));
```

Comment:
❓ I believe the second test should be `isBefore`.

1. When `cacheDuration == null`, we short-circuit out of the conditional with a `true` which is then toggled to a `false`: implying _Not Expired_ ✅ 
2. When `cacheDuration != null`, we continue to the next test; if `now isAfter (created + cacheDuration)` we return `true` which is then toggled to a `false`: implying _Not Expired_ ❌ 
3. When `cacheDuration != null`, we continue to the next test; if `now isNotAfter (created + cacheDuration)` we return `false` which is then toggled to a `true`: implying _Expired_ ❌ 

Unless I'm just misreading this test, it should instead be
```java
return !(cacheDuration == null || created.plus(cacheDuration).isAfter(Instant.now()));
```

Reply:
Honestly, I think this is a case where _not_ simplifying/collapsing conditional tests and being more verbose might be in order. Explicit test branches are easy to reason about.
```java
if (cacheDuration == null) {
  return false;
}

return Instant.now().isAfter(created.plus(cacheDuration));
```
Just a little more explicit and verbose, but I think a lot clearer for a human to read.